In [1]:
import scanpy as sc
import pandas as pd

h5ad_path = r"../data/discovery/taurus_v3/myeloid_final.h5ad"
paired_path = r"../data/discovery/taurus_v3/paired_sample_list.csv"

adata = sc.read_h5ad(h5ad_path)
paired = pd.read_csv(paired_path)

print("adata.shape:", adata.shape)
print("obs cols:", len(adata.obs.columns))
display(adata.obs.head())
print("paired.shape:", paired.shape)
display(paired.head())


adata.shape: (30858, 33075)
obs cols: 39


sample_id Patient Disease    Site Treatment  \
AAAGGATTCGACCATA-1-CID003352-2  CID003352-2     UC1      UC  Rectum      Post   
AAAGGTAGTTCCTAAG-1-CID003352-2  CID003352-2     UC1      UC  Rectum      Post   
AACAAGAAGCGTGAAC-1-CID003352-2  CID003352-2     UC1      UC  Rectum      Post   
AACCACATCCTTATGT-1-CID003352-2  CID003352-2     UC1      UC  Rectum      Post   
AACCATGAGAGTACCG-1-CID003352-2  CID003352-2     UC1      UC  Rectum      Post   

                                Disease_duration Inflammation  Age Gender  \
AAAGGATTCGACCATA-1-CID003352-2              29.0     Inflamed   26      M   
AAAGGTAGTTCCTAAG-1-CID003352-2              29.0     Inflamed   26      M   
AACAAGAAGCGTGAAC-1-CID003352-2              29.0     Inflamed   26      M   
AACCACATCCTTATGT-1-CID003352-2              29.0     Inflamed   26      M   
AACCATGAGAGTACCG-1-CID003352-2              29.0     Inflamed   26      M   

                                Ethnicity  ...  G2M_score phase  \
AAAGGATTCGACCATA-1-CID003352-2  Caucasian  ...  -1.146853    G1   
AAAGGTAGTTCCTAAG-1-CID003352-2  Caucasian  ...  -0.348485    G1   
AACAAGAAGCGTGAAC-1-CID003352-2  Caucasian  ...  -1.851399    G1   
AACCACATCCTTATGT-1-CID003352-2  Caucasian  ...  -0.404429    G1   
AACCATGAGAGTACCG-1-CID003352-2  Caucasian  ...  -1.075175    G1   

                                                   cellbarcode  diff  \
AAAGGATTCGACCATA-1-CID003352-2  AAAGGATTCGACCATA-1-CID003352-2   8.0   
AAAGGTAGTTCCTAAG-1-CID003352-2  AAAGGTAGTTCCTAAG-1-CID003352-2   8.0   
AACAAGAAGCGTGAAC-1-CID003352-2  AACAAGAAGCGTGAAC-1-CID003352-2   8.0   
AACCACATCCTTATGT-1-CID003352-2  AACCACATCCTTATGT-1-CID003352-2   8.0   
AACCATGAGAGTACCG-1-CID003352-2  AACCATGAGAGTACCG-1-CID003352-2   8.0   

                                    final_analysis  minor       major  \
AAAGGATTCGACCATA-1-CID003352-2  C1Qhi IL1Blo macro  Macro  Mono_macro   
AAAGGTAGTTCCTAAG-1-CID003352-2    S100A8 A9hi mono   Mono  Mono_macro   
AACAAGAAGCGTGAAC-1-CID003352-2           CD1Chi DC     DC          DC   
AACCACATCCTTATGT-1-CID003352-2           CD1Chi DC     DC          DC   
AACCATGAGAGTACCG-1-CID003352-2  C1Qhi IL1Bhi macro  Macro  Mono_macro   

                                   sub_bucket   bucket  Remission_status  
AAAGGATTCGACCATA-1-CID003352-2  Mono_macro_DC  Myeloid     Non_Remission  
AAAGGTAGTTCCTAAG-1-CID003352-2  Mono_macro_DC  Myeloid     Non_Remission  
AACAAGAAGCGTGAAC-1-CID003352-2  Mono_macro_DC  Myeloid     Non_Remission  
AACCACATCCTTATGT-1-CID003352-2  Mono_macro_DC  Myeloid     Non_Remission  
AACCATGAGAGTACCG-1-CID003352-2  Mono_macro_DC  Myeloid     Non_Remission  

[5 rows x 39 columns]

paired.shape: (110, 1)


sample_id\tCategory
0  CID003377-1\tCD_Non_Remission
1  CID005757-1\tCD_Non_Remission
2  CID005761-1\tCD_Non_Remission
3  CID003675-1\tCD_Non_Remission
4  CID003674-1\tCD_Non_Remission

In [2]:
with open(paired_path, "r", encoding="utf-8") as f:
    first_line = f.readline()

print("FIRST LINE repr:", repr(first_line))


FIRST LINE repr: 'sample_id\tCategory\n'


In [3]:
paired = pd.read_csv(paired_path, sep="\t")

print("paired.shape:", paired.shape)
print("paired columns:", list(paired.columns))
display(paired.head())


paired.shape: (110, 2)
paired columns: ['sample_id', 'Category']


sample_id          Category
0  CID003377-1  CD_Non_Remission
1  CID005757-1  CD_Non_Remission
2  CID005761-1  CD_Non_Remission
3  CID003675-1  CD_Non_Remission
4  CID003674-1  CD_Non_Remission

In [4]:
# 1) How many unique IDs are in each candidate column?
candidates = ["sample_id", "Patient", "cellbarcode", "Remission_status", "Treatment", "Site", "Disease"]
for c in candidates:
    if c in adata.obs.columns:
        print(c, "nunique =", adata.obs[c].nunique())

# 2) Look at the 3 ID columns side-by-side for a few rows
cols_show = [c for c in ["sample_id", "Patient", "cellbarcode"] if c in adata.obs.columns]
display(adata.obs[cols_show].head())

# 3) For the first 5 Patient IDs, show the distinct metadata patterns
meta = [c for c in ["Patient","Disease","Site","Treatment","Disease_duration","Inflammation","Remission_status"] if c in adata.obs.columns]
tmp = adata.obs[meta].drop_duplicates()
display(tmp.head(20))


sample_id nunique = 216
Patient nunique = 41
cellbarcode nunique = 30858
Remission_status nunique = 4
Treatment nunique = 2
Site nunique = 5
Disease nunique = 3


sample_id Patient  \
AAAGGATTCGACCATA-1-CID003352-2  CID003352-2     UC1   
AAAGGTAGTTCCTAAG-1-CID003352-2  CID003352-2     UC1   
AACAAGAAGCGTGAAC-1-CID003352-2  CID003352-2     UC1   
AACCACATCCTTATGT-1-CID003352-2  CID003352-2     UC1   
AACCATGAGAGTACCG-1-CID003352-2  CID003352-2     UC1   

                                                   cellbarcode  
AAAGGATTCGACCATA-1-CID003352-2  AAAGGATTCGACCATA-1-CID003352-2  
AAAGGTAGTTCCTAAG-1-CID003352-2  AAAGGTAGTTCCTAAG-1-CID003352-2  
AACAAGAAGCGTGAAC-1-CID003352-2  AACAAGAAGCGTGAAC-1-CID003352-2  
AACCACATCCTTATGT-1-CID003352-2  AACCACATCCTTATGT-1-CID003352-2  
AACCATGAGAGTACCG-1-CID003352-2  AACCATGAGAGTACCG-1-CID003352-2

Patient Disease              Site Treatment  \
AAAGGATTCGACCATA-1-CID003352-2     UC1      UC            Rectum      Post   
AACAACCTCCCGGTAG-1-CID003353-1     UC1      UC  Descending_Colon      Post   
ACCAACACATATGCGT-1-CID003354-1     UC1      UC   Ascending_Colon      Post   
AAAGGTATCTAACACG-1-CID003355-1     UC1      UC    Terminal_Ileum      Post   
AAAGGGCCAGTAGATA-1-CID003356-1     UC1      UC            Rectum       Pre   
AAACGAAGTCCAATCA-1-CID003357-1     UC1      UC    Terminal_Ileum       Pre   
AACAGGGCAGCACACC-1-CID003359-1     UC1      UC   Ascending_Colon       Pre   
AAACCCACAGCACAAG-1-CID003360-1     UC2      UC            Rectum      Post   
AAACCCACACCCTATC-1-CID003361-1     UC2      UC           Sigmoid      Post   
AAAGGATAGCATTTGC-1-CID003362-1     UC2      UC  Descending_Colon      Post   
AAAGAACAGCGATGAC-1-CID003363-1     UC2      UC            Rectum       Pre   
AAACGCTAGCCTCTTC-1-CID003364-1     UC2      UC           Sigmoid       Pre   
AAGTGAAGTGCTCGTG-1-CID003365-1     UC2      UC  Descending_Colon       Pre   
AAAGAACTCTTCTTCC-1-CID003366-1     UC3      UC            Rectum      Post   
AAAGGGCCACTCCCTA-1-CID003367-2     UC3      UC  Descending_Colon      Post   
AAAGGATGTGAGGCAT-1-CID003368-1     UC3      UC   Ascending_Colon      Post   
AAACCCAAGATCCGAG-1-CID003369-1     UC3      UC    Terminal_Ileum      Post   
AAACGCTAGAGATGCC-1-CID003370-1     UC3      UC            Rectum       Pre   
AAAGTCCTCGATTGAC-1-CID003371-2     UC3      UC  Descending_Colon       Pre   
AAACGCTAGGGCCCTT-1-CID003372-1     UC3      UC   Ascending_Colon       Pre   

                                Disease_duration  Inflammation  \
AAAGGATTCGACCATA-1-CID003352-2              29.0      Inflamed   
AACAACCTCCCGGTAG-1-CID003353-1              29.0  Non_Inflamed   
ACCAACACATATGCGT-1-CID003354-1              29.0  Non_Inflamed   
AAAGGTATCTAACACG-1-CID003355-1              29.0  Non_Inflamed   
AAAGGGCCAGTAGATA-1-CID003356-1              29.0      Inflamed   
AAACGAAGTCCAATCA-1-CID003357-1              29.0  Non_Inflamed   
AACAGGGCAGCACACC-1-CID003359-1              29.0  Non_Inflamed   
AAACCCACAGCACAAG-1-CID003360-1             146.0      Inflamed   
AAACCCACACCCTATC-1-CID003361-1             146.0      Inflamed   
AAAGGATAGCATTTGC-1-CID003362-1             146.0      Inflamed   
AAAGAACAGCGATGAC-1-CID003363-1             146.0      Inflamed   
AAACGCTAGCCTCTTC-1-CID003364-1             146.0      Inflamed   
AAGTGAAGTGCTCGTG-1-CID003365-1             146.0      Inflamed   
AAAGAACTCTTCTTCC-1-CID003366-1             287.0      Inflamed   
AAAGGGCCACTCCCTA-1-CID003367-2             287.0      Inflamed   
AAAGGATGTGAGGCAT-1-CID003368-1             287.0      Inflamed   
AAACCCAAGATCCGAG-1-CID003369-1             287.0  Non_Inflamed   
AAACGCTAGAGATGCC-1-CID003370-1             287.0  Non_Inflamed   
AAAGTCCTCGATTGAC-1-CID003371-2             287.0  Non_Inflamed   
AAACGCTAGGGCCCTT-1-CID003372-1             287.0      Inflamed   

                               Remission_status  
AAAGGATTCGACCATA-1-CID003352-2    Non_Remission  
AACAACCTCCCGGTAG-1-CID003353-1    Non_Remission  
ACCAACACATATGCGT-1-CID003354-1    Non_Remission  
AAAGGTATCTAACACG-1-CID003355-1    Non_Remission  
AAAGGGCCAGTAGATA-1-CID003356-1    Non_Remission  
AAACGAAGTCCAATCA-1-CID003357-1    Non_Remission  
AACAGGGCAGCACACC-1-CID003359-1    Non_Remission  
AAACCCACAGCACAAG-1-CID003360-1    Non_Remission  
AAACCCACACCCTATC-1-CID003361-1    Non_Remission  
AAAGGATAGCATTTGC-1-CID003362-1    Non_Remission  
AAAGAACAGCGATGAC-1-CID003363-1    Non_Remission  
AAACGCTAGCCTCTTC-1-CID003364-1    Non_Remission  
AAGTGAAGTGCTCGTG-1-CID003365-1    Non_Remission  
AAAGAACTCTTCTTCC-1-CID003366-1    Non_Remission  
AAAGGGCCACTCCCTA-1-CID003367-2    Non_Remission  
AAAGGATGTGAGGCAT-1-CID003368-1    Non_Remission  
AAACCCAAGATCCGAG-1-CID003369-1    Non_Remission  
AAACGCTAGAGATGCC-1-CID003370-1    Non_Remission  
AAAGTCCTCGATTGAC-1-CID003371-2    Non_Remission  
AAACG

In [5]:
# Build a sample-level table from the myeloid object
sample_cols = [
    "Patient",          # biopsy/sample id (we will rename to sample_id)
    "Disease",          # UC / CD / ?
    "Site",             # Rectum, etc.
    "Treatment",        # Pre / Post
    "Remission_status", # Remission / Non_Remission (and possibly others)
    "Inflammation",     # Inflamed / Non_Inflamed
    "Age",
    "Gender",
    "Ethnicity",
]

# Keep only columns that exist
sample_cols = [c for c in sample_cols if c in adata.obs.columns]

samples = (
    adata.obs[sample_cols]
    .drop_duplicates()
    .rename(columns={
        "Patient": "sample_id",
        "Treatment": "timepoint",
        "Remission_status": "response",
    })
    .copy()
)

# Add subject_id by removing the trailing "-1"/"-2" from CIDxxxxx-1
samples["subject_id"] = samples["sample_id"].astype(str).str.replace(r"-\d+$", "", regex=True)

# Add cell counts per sample (myeloid cells only, since object is myeloid)
cell_counts = adata.obs.groupby("Patient").size().rename("n_myeloid_cells").reset_index().rename(columns={"Patient":"sample_id"})
samples = samples.merge(cell_counts, on="sample_id", how="left")

# Sort for readability
samples = samples.sort_values(["subject_id","sample_id"]).reset_index(drop=True)

print("samples.shape:", samples.shape)
display(samples.head(10))

# Save to disk (repo local; your .gitignore doesn't ignore results/tables, so we write under data/)
out_path = r"../data/discovery/taurus_v3/samples.tsv"
samples.to_csv(out_path, sep="\t", index=False)
print("Wrote:", out_path)


samples.shape: (216, 11)


sample_id Disease              Site timepoint       response  Inflammation  \
0       CD1      CD            Rectum      Post  Non_Remission  Non_Inflamed   
1       CD1      CD  Descending_Colon      Post  Non_Remission  Non_Inflamed   
2       CD1      CD   Ascending_Colon      Post  Non_Remission  Non_Inflamed   
3       CD1      CD    Terminal_Ileum      Post  Non_Remission      Inflamed   
4       CD1      CD            Rectum       Pre  Non_Remission  Non_Inflamed   
5       CD1      CD  Descending_Colon       Pre  Non_Remission  Non_Inflamed   
6       CD1      CD   Ascending_Colon       Pre  Non_Remission  Non_Inflamed   
7       CD1      CD    Terminal_Ileum       Pre  Non_Remission      Inflamed   
8      CD10      CD    Terminal_Ileum       Pre      Remission      Inflamed   
9      CD10      CD   Ascending_Colon       Pre      Remission      Inflamed   

   Age Gender  Ethnicity subject_id  n_myeloid_cells  
0   28      F  Caucasian        CD1             1946  
1   28      F  Caucasian        CD1             1946  
2   28      F  Caucasian        CD1             1946  
3   28      F  Caucasian        CD1             1946  
4   28      F  Caucasian        CD1             1946  
5   28      F  Caucasian        CD1             1946  
6   28      F  Caucasian        CD1             1946  
7   28      F  Caucasian        CD1             1946  
8   27      M  Caucasian       CD10              869  
9   27      M  Caucasian       CD10              869

Wrote: ../data/discovery/taurus_v3/samples.tsv


In [6]:
print("obs_names example:", list(adata.obs_names[:3]))
print("obs_names unique:", adata.obs_names.nunique())

print("\nadata.obs['sample_id'] head:", list(adata.obs["sample_id"].head()))
print("sample_id nunique:", adata.obs["sample_id"].nunique())

print("\nadata.obs['Patient'] head:", list(adata.obs["Patient"].head()))
print("Patient nunique:", adata.obs["Patient"].nunique())


obs_names example: ['AAAGGATTCGACCATA-1-CID003352-2', 'AAAGGTAGTTCCTAAG-1-CID003352-2', 'AACAAGAAGCGTGAAC-1-CID003352-2']
obs_names unique: 30858

adata.obs['sample_id'] head: ['CID003352-2', 'CID003352-2', 'CID003352-2', 'CID003352-2', 'CID003352-2']
sample_id nunique: 216

adata.obs['Patient'] head: ['UC1', 'UC1', 'UC1', 'UC1', 'UC1']
Patient nunique: 41


In [7]:
# Rebuild sample-level table correctly: one row per biopsy/sample_id
sample_cols = [
    "sample_id",        # biopsy/sample id (CIDxxxxx-1/2)
    "Patient",          # subject label (UC1/CD10/etc)
    "Disease",          # UC / CD / ?
    "Site",             # Rectum, etc
    "Treatment",        # Pre / Post
    "Remission_status", # Remission / Non_Remission etc
    "Inflammation",
    "Age",
    "Gender",
    "Ethnicity",
]

sample_cols = [c for c in sample_cols if c in adata.obs.columns]

samples = (
    adata.obs[sample_cols]
    .drop_duplicates()
    .rename(columns={
        "Patient": "subject_id",
        "Treatment": "timepoint",
        "Remission_status": "response",
    })
    .copy()
)

# Myeloid cell counts per biopsy
cell_counts = adata.obs.groupby("sample_id").size().rename("n_myeloid_cells").reset_index()
samples = samples.merge(cell_counts, on="sample_id", how="left")

# Sanity check: should be 216 rows and sample_id unique
print("samples rows:", len(samples))
print("sample_id unique:", samples["sample_id"].nunique())

display(samples.head(10))

out_path = r"../data/discovery/taurus_v3/samples.tsv"
samples.to_csv(out_path, sep="\t", index=False)
print("Wrote:", out_path)


samples rows: 216
sample_id unique: 216


sample_id subject_id Disease              Site timepoint       response  \
0  CID003352-2        UC1      UC            Rectum      Post  Non_Remission   
1  CID003353-1        UC1      UC  Descending_Colon      Post  Non_Remission   
2  CID003354-1        UC1      UC   Ascending_Colon      Post  Non_Remission   
3  CID003355-1        UC1      UC    Terminal_Ileum      Post  Non_Remission   
4  CID003356-1        UC1      UC            Rectum       Pre  Non_Remission   
5  CID003357-1        UC1      UC    Terminal_Ileum       Pre  Non_Remission   
6  CID003359-1        UC1      UC   Ascending_Colon       Pre  Non_Remission   
7  CID003360-1        UC2      UC            Rectum      Post  Non_Remission   
8  CID003361-1        UC2      UC           Sigmoid      Post  Non_Remission   
9  CID003362-1        UC2      UC  Descending_Colon      Post  Non_Remission   

   Inflammation  Age Gender  Ethnicity  n_myeloid_cells  
0      Inflamed   26      M  Caucasian              258  
1  Non_Inflamed   26      M  Caucasian               36  
2  Non_Inflamed   26      M  Caucasian               34  
3  Non_Inflamed   26      M  Caucasian              103  
4      Inflamed   26      M  Caucasian               84  
5  Non_Inflamed   26      M  Caucasian              141  
6  Non_Inflamed   26      M  Caucasian               24  
7      Inflamed   31      F  Caucasian              155  
8      Inflamed   31      F  Caucasian               90  
9      Inflamed   31      F  Caucasian               57

Wrote: ../data/discovery/taurus_v3/samples.tsv


In [8]:
# Join paired category onto our samples table
samples2 = samples.merge(paired, on="sample_id", how="left")

print("Total samples:", len(samples2))
print("Samples with Category (in paired list):", samples2["Category"].notna().sum())

# Show the Category distribution among the matched samples
print("\nCategory counts:")
print(samples2["Category"].value_counts(dropna=True))

# Quick check: how many paired IDs are missing from our h5ad samples?
missing_from_h5ad = set(paired["sample_id"]) - set(samples2["sample_id"])
print("\nPaired sample_ids missing from this myeloid h5ad:", len(missing_from_h5ad))
print("Example missing IDs:", list(sorted(missing_from_h5ad))[:10])

display(samples2.head(10))


Total samples: 216
Samples with Category (in paired list): 110

Category counts:
Category
UC_Non_Remission    42
CD_Remission        38
UC_Remission        16
CD_Non_Remission    14
Name: count, dtype: int64

Paired sample_ids missing from this myeloid h5ad: 0
Example missing IDs: []


sample_id subject_id Disease              Site timepoint       response  \
0  CID003352-2        UC1      UC            Rectum      Post  Non_Remission   
1  CID003353-1        UC1      UC  Descending_Colon      Post  Non_Remission   
2  CID003354-1        UC1      UC   Ascending_Colon      Post  Non_Remission   
3  CID003355-1        UC1      UC    Terminal_Ileum      Post  Non_Remission   
4  CID003356-1        UC1      UC            Rectum       Pre  Non_Remission   
5  CID003357-1        UC1      UC    Terminal_Ileum       Pre  Non_Remission   
6  CID003359-1        UC1      UC   Ascending_Colon       Pre  Non_Remission   
7  CID003360-1        UC2      UC            Rectum      Post  Non_Remission   
8  CID003361-1        UC2      UC           Sigmoid      Post  Non_Remission   
9  CID003362-1        UC2      UC  Descending_Colon      Post  Non_Remission   

   Inflammation  Age Gender  Ethnicity  n_myeloid_cells          Category  
0      Inflamed   26      M  Caucasian              258  UC_Non_Remission  
1  Non_Inflamed   26      M  Caucasian               36               NaN  
2  Non_Inflamed   26      M  Caucasian               34               NaN  
3  Non_Inflamed   26      M  Caucasian              103               NaN  
4      Inflamed   26      M  Caucasian               84  UC_Non_Remission  
5  Non_Inflamed   26      M  Caucasian              141               NaN  
6  Non_Inflamed   26      M  Caucasian               24               NaN  
7      Inflamed   31      F  Caucasian              155  UC_Non_Remission  
8      Inflamed   31      F  Caucasian               90  UC_Non_Remission  
9      Inflamed   31      F  Caucasian               57               NaN

In [10]:
import os, sys
from pathlib import Path

# Current notebook folder should be .../Single-Cell-Reanalysis/notebooks
print("cwd:", os.getcwd())

# Add repo root (parent of notebooks/) to sys.path
repo_root = Path(os.getcwd()).parent
sys.path.insert(0, str(repo_root))

print("repo_root added to path:", repo_root)
print("src exists?:", (repo_root / "src").exists())


cwd: C:\Users\Jaeho\OneDrive\Desktop\GitHub\Single-Cell-Reanalysis\notebooks
repo_root added to path: C:\Users\Jaeho\OneDrive\Desktop\GitHub\Single-Cell-Reanalysis
src exists?: True


In [11]:
from src.state_scoring import load_gmt, score_modules
print("imported state_scoring OK")


imported state_scoring OK


In [12]:
gmt_path = r"../src/gene_sets/hallmark_selected.gmt"
gene_sets = load_gmt(gmt_path)

modules = [
    "HALLMARK_OXIDATIVE_PHOSPHORYLATION",
    "HALLMARK_REACTIVE_OXYGEN_SPECIES_PATHWAY",
    "HALLMARK_TNFA_SIGNALING_VIA_NFKB",
    "HALLMARK_INFLAMMATORY_RESPONSE",
    "HALLMARK_INTERFERON_ALPHA_RESPONSE",
    "HALLMARK_INTERFERON_GAMMA_RESPONSE",
]

used_sets = {m: gene_sets[m] for m in modules if m in gene_sets}
print("Loaded gene sets:", list(used_sets.keys()))

adata = score_modules(adata, used_sets)

# Find which columns got added (your function may name them differently)
hallmark_cols = [c for c in adata.obs.columns if "HALLMARK_" in c]
print("Columns containing 'HALLMARK_':", hallmark_cols[:30])
print("Total hallmark-like columns:", len(hallmark_cols))

# Show a preview of the first few hallmark columns
display(adata.obs[hallmark_cols[:6]].head())


Loaded gene sets: ['HALLMARK_OXIDATIVE_PHOSPHORYLATION', 'HALLMARK_REACTIVE_OXYGEN_SPECIES_PATHWAY', 'HALLMARK_TNFA_SIGNALING_VIA_NFKB', 'HALLMARK_INFLAMMATORY_RESPONSE', 'HALLMARK_INTERFERON_ALPHA_RESPONSE', 'HALLMARK_INTERFERON_GAMMA_RESPONSE']


AttributeError: 'NoneType' object has no attribute 'obs'

In [9]:
from src.state_scoring import load_gmt, score_modules  # should already exist in your repo

gmt_path = r"../src/gene_sets/hallmark_selected.gmt"
gene_sets = load_gmt(gmt_path)

modules = [
    "HALLMARK_OXIDATIVE_PHOSPHORYLATION",
    "HALLMARK_REACTIVE_OXYGEN_SPECIES_PATHWAY",
    "HALLMARK_TNFA_SIGNALING_VIA_NFKB",
    "HALLMARK_INFLAMMATORY_RESPONSE",
    "HALLMARK_INTERFERON_ALPHA_RESPONSE",
    "HALLMARK_INTERFERON_GAMMA_RESPONSE",
]

# Keep only the sets that are actually found in the GMT
used_sets = {m: gene_sets[m] for m in modules if m in gene_sets}
print("Loaded gene sets:", list(used_sets.keys()))

adata = score_modules(adata, used_sets)  # this should add score columns to adata.obs

# Show the new score columns (first 10 rows for a couple)
new_cols = [c for c in adata.obs.columns if "HALLMARK_" in c]
print("HALLMARK score cols found in adata.obs:", [c for c in new_cols][:20])
display(adata.obs[[c for c in new_cols][:6]].head())


ModuleNotFoundError: No module named 'src'

In [13]:
import scanpy as sc

# 1) Reload adata (because it got overwritten to None)
adata = sc.read_h5ad(h5ad_path)
print("Reloaded adata.shape:", adata.shape)

# 2) Score modules WITHOUT overwriting adata
before_cols = set(adata.obs.columns)

ret = score_modules(adata, used_sets)  # likely returns None; modifies adata in-place
print("score_modules return value:", ret)

after_cols = set(adata.obs.columns)
new_cols = sorted(list(after_cols - before_cols))

print("New obs columns added:", new_cols)
print("Count new columns:", len(new_cols))

# quick preview if any new columns exist
if len(new_cols) > 0:
    display(adata.obs[new_cols[:6]].head())


Reloaded adata.shape: (30858, 33075)
score_modules return value: None
New obs columns added: ['score_HALLMARK_INFLAMMATORY_RESPONSE', 'score_HALLMARK_INTERFERON_ALPHA_RESPONSE', 'score_HALLMARK_INTERFERON_GAMMA_RESPONSE', 'score_HALLMARK_OXIDATIVE_PHOSPHORYLATION', 'score_HALLMARK_REACTIVE_OXYGEN_SPECIES_PATHWAY', 'score_HALLMARK_TNFA_SIGNALING_VIA_NFKB']
Count new columns: 6


score_HALLMARK_INFLAMMATORY_RESPONSE  \
AAAGGATTCGACCATA-1-CID003352-2                              0.619270   
AAAGGTAGTTCCTAAG-1-CID003352-2                              0.520056   
AACAAGAAGCGTGAAC-1-CID003352-2                              0.457865   
AACCACATCCTTATGT-1-CID003352-2                              1.245955   
AACCATGAGAGTACCG-1-CID003352-2                              0.539494   

                                score_HALLMARK_INTERFERON_ALPHA_RESPONSE  \
AAAGGATTCGACCATA-1-CID003352-2                                  3.003765   
AAAGGTAGTTCCTAAG-1-CID003352-2                                  0.766190   
AACAAGAAGCGTGAAC-1-CID003352-2                                  2.598017   
AACCACATCCTTATGT-1-CID003352-2                                  1.124247   
AACCATGAGAGTACCG-1-CID003352-2                                  1.820783   

                                score_HALLMARK_INTERFERON_GAMMA_RESPONSE  \
AAAGGATTCGACCATA-1-CID003352-2                                  2.790739   
AAAGGTAGTTCCTAAG-1-CID003352-2                                  0.825017   
AACAAGAAGCGTGAAC-1-CID003352-2                                  3.051646   
AACCACATCCTTATGT-1-CID003352-2                                  1.641597   
AACCATGAGAGTACCG-1-CID003352-2                                  1.939697   

                                score_HALLMARK_OXIDATIVE_PHOSPHORYLATION  \
AAAGGATTCGACCATA-1-CID003352-2                                  1.710587   
AAAGGTAGTTCCTAAG-1-CID003352-2                                  0.213036   
AACAAGAAGCGTGAAC-1-CID003352-2                                  1.569372   
AACCACATCCTTATGT-1-CID003352-2                                 -0.027733   
AACCATGAGAGTACCG-1-CID003352-2                                  0.867895   

                                score_HALLMARK_REACTIVE_OXYGEN_SPECIES_PATHWAY  \
AAAGGATTCGACCATA-1-CID003352-2                                       11.702114   
AAAGGTAGTTCCTAAG-1-CID003352-2                                        0.666734   
AACAAGAAGCGTGAAC-1-CID003352-2                                        1.211425   
AACCACATCCTTATGT-1-CID003352-2                                        1.621658   
AACCATGAGAGTACCG-1-CID003352-2                                       11.098024   

                                score_HALLMARK_TNFA_SIGNALING_VIA_NFKB  
AAAGGATTCGACCATA-1-CID003352-2                                1.926647  
AAAGGTAGTTCCTAAG-1-CID003352-2                                2.512170  
AACAAGAAGCGTGAAC-1-CID003352-2                                1.375023  
AACCACATCCTTATGT-1-CID003352-2                                3.966268  
AACCATGAGAGTACCG-1-CID003352-2                                5.755377

In [14]:
from src.state_scoring import assign_state_high

score_cols = [
    "score_HALLMARK_OXIDATIVE_PHOSPHORYLATION",
    "score_HALLMARK_REACTIVE_OXYGEN_SPECIES_PATHWAY",
    "score_HALLMARK_TNFA_SIGNALING_VIA_NFKB",
    "score_HALLMARK_INFLAMMATORY_RESPONSE",
    "score_HALLMARK_INTERFERON_ALPHA_RESPONSE",
    "score_HALLMARK_INTERFERON_GAMMA_RESPONSE",
]

# This should add boolean columns like state_high_<scorecol> or similar depending on your function
ret = assign_state_high(
    adata,
    score_cols=score_cols,
    groupby="sample_id",
    top_frac=0.20,
)
print("assign_state_high return value:", ret)

# Find new columns added
added = [c for c in adata.obs.columns if "state_high" in c]
print("state_high-like columns:", added)

# Quick sanity: within each sample, percent high should be ~20% per module
import pandas as pd

for sc_col in score_cols:
    # guess the output column name
    candidates = [c for c in added if sc_col in c] + [f"state_high_{sc_col}", f"state_high_{sc_col.replace('score_','')}"]
    high_col = next((c for c in candidates if c in adata.obs.columns), None)

    if high_col is None:
        print("Could not find state-high column for:", sc_col)
        continue

    frac = adata.obs.groupby("sample_id")[high_col].mean()
    print(sc_col, "  per-sample high% median=", float(frac.median()*100), " min=", float(frac.min()*100), " max=", float(frac.max()*100))


TypeError: assign_state_high() got an unexpected keyword argument 'score_cols'

In [15]:
import inspect
from src.state_scoring import assign_state_high

print(inspect.signature(assign_state_high))


(adata: 'AnnData', score_col: 'str', groupby: 'str' = 'sample_id', q: 'float' = 0.8, out_col: 'str | None' = None) -> 'None'


In [16]:
score_cols = [
    "score_HALLMARK_OXIDATIVE_PHOSPHORYLATION",
    "score_HALLMARK_REACTIVE_OXYGEN_SPECIES_PATHWAY",
    "score_HALLMARK_TNFA_SIGNALING_VIA_NFKB",
    "score_HALLMARK_INFLAMMATORY_RESPONSE",
    "score_HALLMARK_INTERFERON_ALPHA_RESPONSE",
    "score_HALLMARK_INTERFERON_GAMMA_RESPONSE",
]

# Create state_high_* columns (top 20% within each sample_id)
for sc_col in score_cols:
    out_col = "state_high_" + sc_col.replace("score_", "")
    assign_state_high(
        adata,
        score_col=sc_col,
        groupby="sample_id",
        q=0.8,          # top 20%
        out_col=out_col
    )

state_cols = [c for c in adata.obs.columns if c.startswith("state_high_")]
print("Created state_high columns:", state_cols)

# Sanity check: per sample, should be ~20% high
for sc_col in score_cols:
    out_col = "state_high_" + sc_col.replace("score_", "")
    frac = adata.obs.groupby("sample_id")[out_col].mean()
    print(out_col, "median%=", float(frac.median()*100), "min%=", float(frac.min()*100), "max%=", float(frac.max()*100))


Created state_high columns: ['state_high_HALLMARK_OXIDATIVE_PHOSPHORYLATION', 'state_high_HALLMARK_REACTIVE_OXYGEN_SPECIES_PATHWAY', 'state_high_HALLMARK_TNFA_SIGNALING_VIA_NFKB', 'state_high_HALLMARK_INFLAMMATORY_RESPONSE', 'state_high_HALLMARK_INTERFERON_ALPHA_RESPONSE', 'state_high_HALLMARK_INTERFERON_GAMMA_RESPONSE']
state_high_HALLMARK_OXIDATIVE_PHOSPHORYLATION median%= 20.32967032967033 min%= 20.0 max%= 23.52941176470588
state_high_HALLMARK_REACTIVE_OXYGEN_SPECIES_PATHWAY median%= 20.327436790851426 min%= 20.0 max%= 23.52941176470588
state_high_HALLMARK_TNFA_SIGNALING_VIA_NFKB median%= 20.327436790851426 min%= 20.0 max%= 23.52941176470588
state_high_HALLMARK_INFLAMMATORY_RESPONSE median%= 20.327436790851426 min%= 20.0 max%= 23.52941176470588
state_high_HALLMARK_INTERFERON_ALPHA_RESPONSE median%= 20.327436790851426 min%= 20.0 max%= 23.52941176470588
state_high_HALLMARK_INTERFERON_GAMMA_RESPONSE median%= 20.327436790851426 min%= 20.0 max%= 23.52941176470588


In [17]:
import pandas as pd

# 1) Start from sample metadata table you already built earlier:
# samples = one row per biopsy/sample_id (216 rows)
# paired = 110 rows with Category

samples2 = samples.merge(paired, on="sample_id", how="left")

# Keep only paired samples
paired_samples = samples2[samples2["Category"].notna()].copy()
print("paired_samples rows:", len(paired_samples), " unique sample_id:", paired_samples["sample_id"].nunique())

# 2) Compute per-sample %high and mean scores from adata.obs
rows = []
for sid, idx in adata.obs.groupby("sample_id").groups.items():
    # only keep paired sample_ids
    if sid not in set(paired_samples["sample_id"]):
        continue
    sub = adata.obs.loc[idx]

    row = {"sample_id": sid, "n_myeloid_cells": int(sub.shape[0])}

    # mean scores
    for sc_col in score_cols:
        row[f"mean_{sc_col}"] = float(sub[sc_col].mean())

    # % state-high
    for sc_col in score_cols:
        st_col = "state_high_" + sc_col.replace("score_", "")
        row[f"pct_high_{st_col.replace('state_high_','')}"] = float(sub[st_col].mean() * 100.0)

    rows.append(row)

state_summary = pd.DataFrame(rows)

# 3) Join sample metadata + paired Category
paired_state = paired_samples.merge(state_summary, on="sample_id", how="left")

print("paired_state shape:", paired_state.shape)
display(paired_state.head(5))

# Save locally
out_path = r"../data/discovery/taurus_v3/paired_state_summary.tsv"
paired_state.to_csv(out_path, sep="\t", index=False)
print("Wrote:", out_path)


paired_samples rows: 110  unique sample_id: 110
paired_state shape: (110, 25)


sample_id subject_id Disease     Site timepoint       response  \
0  CID003352-2        UC1      UC   Rectum      Post  Non_Remission   
1  CID003356-1        UC1      UC   Rectum       Pre  Non_Remission   
2  CID003360-1        UC2      UC   Rectum      Post  Non_Remission   
3  CID003361-1        UC2      UC  Sigmoid      Post  Non_Remission   
4  CID003363-1        UC2      UC   Rectum       Pre  Non_Remission   

  Inflammation  Age Gender  Ethnicity  ...  \
0     Inflamed   26      M  Caucasian  ...   
1     Inflamed   26      M  Caucasian  ...   
2     Inflamed   31      F  Caucasian  ...   
3     Inflamed   31      F  Caucasian  ...   
4     Inflamed   31      F  Caucasian  ...   

   mean_score_HALLMARK_TNFA_SIGNALING_VIA_NFKB  \
0                                     1.947804   
1                                     1.510042   
2                                     1.022152   
3                                     1.467810   
4                                     1.776840   

  mean_score_HALLMARK_INFLAMMATORY_RESPONSE  \
0                                  0.583588   
1                                  0.575124   
2                                  0.291921   
3                                  0.446541   
4                                  0.653824   

   mean_score_HALLMARK_INTERFERON_ALPHA_RESPONSE  \
0                                       1.269857   
1                                       0.909192   
2                                       0.779126   
3                                       1.384615   
4                                       1.036769   

   mean_score_HALLMARK_INTERFERON_GAMMA_RESPONSE  \
0                                       1.293405   
1                                       1.171730   
2                                       0.845704   
3                                       1.429119   
4                                       1.160372   

   pct_high_HALLMARK_OXIDATIVE_PHOSPHORYLATION  \
0                                    20.155039   
1                                    20.238095   
2                                    20.000000   
3                                    20.000000   
4                                    20.652174   

   pct_high_HALLMARK_REACTIVE_OXYGEN_SPECIES_PATHWAY  \
0                                          20.155039   
1                                          20.238095   
2                                          20.000000   
3                                          20.000000   
4                                          20.652174   

   pct_high_HALLMARK_TNFA_SIGNALING_VIA_NFKB  \
0                                  20.155039   
1                                  20.238095   
2                                  20.000000   
3                                  20.000000   
4                                  20.652174   

   pct_high_HALLMARK_INFLAMMATORY_RESPONSE  \
0                                20.155039   
1                                20.238095   
2                                20.000000   
3                                20.000000   
4                                20.652174   

   pct_high_HALLMARK_INTERFERON_ALPHA_RESPONSE  \
0                                    20.155039   
1                                    20.238095   
2                                    20.000000   
3                                    20.000000   
4                                    20.652174   

   pct_high_HALLMARK_INTERFERON_GAMMA_RESPONSE  
0                                    20.155039  
1                                    20.238095  
2                                    20.000000  
3                                    20.000000  
4                                    20.652174  

[5 rows x 25 columns]

Wrote: ../data/discovery/taurus_v3/paired_state_summary.tsv


In [18]:
import numpy as np
import pandas as pd

df = paired_state.copy()

# Define response group from Category (CD_Remission, UC_Non_Remission, etc.)
df["resp_group"] = df["Category"].str.contains("Non_Remission").map({True: "Non_Remission", False: "Remission"})

# We will compute delta = Post - Pre per subject_id (within each disease/site combo may have multiple biopsies)
# First check how many Pre/Post per subject
counts = df.pivot_table(index="subject_id", columns="timepoint", values="sample_id", aggfunc="count").fillna(0).astype(int)
print("Subjects with both Pre and Post:", int(((counts.get("Pre",0)>0) & (counts.get("Post",0)>0)).sum()))
display(counts.head(10))

score_mean_cols = [
    "mean_score_HALLMARK_OXIDATIVE_PHOSPHORYLATION",
    "mean_score_HALLMARK_REACTIVE_OXYGEN_SPECIES_PATHWAY",
    "mean_score_HALLMARK_TNFA_SIGNALING_VIA_NFKB",
    "mean_score_HALLMARK_INFLAMMATORY_RESPONSE",
    "mean_score_HALLMARK_INTERFERON_ALPHA_RESPONSE",
    "mean_score_HALLMARK_INTERFERON_GAMMA_RESPONSE",
]

# To keep it simple: average across biopsies per subject per timepoint, then compute Post-Pre
subj_tp = df.groupby(["subject_id","resp_group","timepoint"], as_index=False)[score_mean_cols].mean()

pre = subj_tp[subj_tp["timepoint"]=="Pre"].set_index(["subject_id","resp_group"])
post = subj_tp[subj_tp["timepoint"]=="Post"].set_index(["subject_id","resp_group"])

paired_idx = pre.index.intersection(post.index)
deltas = post.loc[paired_idx, score_mean_cols] - pre.loc[paired_idx, score_mean_cols]
deltas = deltas.reset_index()

print("Paired subjects used:", len(deltas))
display(deltas.head())

# Summarize deltas by response group (median + mean)
summary = deltas.groupby("resp_group")[score_mean_cols].agg(["median","mean","count"])
display(summary)


Subjects with both Pre and Post: 32


timepoint   Post  Pre
subject_id           
CD1            1    1
CD10           2    2
CD11           2    2
CD12           1    1
CD13           4    4
CD14           2    2
CD15           1    1
CD2            1    1
CD3            2    2
CD4            2    2

Paired subjects used: 32


subject_id     resp_group  mean_score_HALLMARK_OXIDATIVE_PHOSPHORYLATION  \
0        CD1  Non_Remission                                      -0.237335   
1       CD10      Remission                                      -0.030364   
2       CD11  Non_Remission                                      -0.168442   
3       CD12  Non_Remission                                       0.003032   
4       CD13      Remission                                      -0.165764   

   mean_score_HALLMARK_REACTIVE_OXYGEN_SPECIES_PATHWAY  \
0                                          -0.446986     
1                                          -0.284630     
2                                          -0.596953     
3                                           0.134318     
4                                          -2.752870     

   mean_score_HALLMARK_TNFA_SIGNALING_VIA_NFKB  \
0                                     0.179057   
1                                    -1.439471   
2                                    -0.859057   
3                                     0.150592   
4                                    -0.913997   

   mean_score_HALLMARK_INFLAMMATORY_RESPONSE  \
0                                   0.680385   
1                                  -0.219837   
2                                  -0.128495   
3                                   0.193617   
4                                  -0.348006   

   mean_score_HALLMARK_INTERFERON_ALPHA_RESPONSE  \
0                                       0.205954   
1                                      -0.090665   
2                                      -0.586845   
3                                       0.263524   
4                                      -0.428279   

   mean_score_HALLMARK_INTERFERON_GAMMA_RESPONSE  
0                                       0.132871  
1                                      -0.184734  
2                                      -0.596444  
3                                       0.294360  
4                                      -0.560874

mean_score_HALLMARK_OXIDATIVE_PHOSPHORYLATION                 \
                                                     median     mean count   
resp_group                                                                   
Non_Remission                                     -0.032072 -0.06020    18   
Remission                                         -0.091295 -0.12376    14   

              mean_score_HALLMARK_REACTIVE_OXYGEN_SPECIES_PATHWAY            \
                                                           median      mean   
resp_group                                                                    
Non_Remission                                          -0.112957  -0.307243   
Remission                                              -0.583639  -1.157662   

                    mean_score_HALLMARK_TNFA_SIGNALING_VIA_NFKB            \
              count                                      median      mean   
resp_group                                                                  
Non_Remission    18                                   -0.078699 -0.152202   
Remission        14                                   -0.353314 -0.311325   

                    mean_score_HALLMARK_INFLAMMATORY_RESPONSE                  \
              count                                    median      mean count   
resp_group                                                                      
Non_Remission    18                                 -0.022633 -0.000241    18   
Remission        14                                 -0.206713 -0.148292    14   

              mean_score_HALLMARK_INTERFERON_ALPHA_RESPONSE                  \
                                                     median      mean count   
resp_group                                                                    
Non_Remission                                      0.028791 -0.049635    18   
Remission                                         -0.299160 -0.357878    14   

              mean_score_HALLMARK_INTERFERON_GAMMA_RESPONSE                  
                                                     median      mean count  
resp_group                                                                   
Non_Remission                                      0.017298 -0.083483    18  
Remission                                         -0.323563 -0.381734    14

In [19]:
# Define combined state: high OXPHOS AND high ROS
adata.obs["state_high_MITO_ROS"] = (
    adata.obs["state_high_HALLMARK_OXIDATIVE_PHOSPHORYLATION"] &
    adata.obs["state_high_HALLMARK_REACTIVE_OXYGEN_SPECIES_PATHWAY"]
)

# Per-sample % of myeloid cells in combined state
combo = (
    adata.obs.groupby("sample_id")["state_high_MITO_ROS"]
    .mean()
    .mul(100)
    .rename("pct_high_MITO_ROS")
    .reset_index()
)

# Join onto paired_state
paired_combo = paired_state.merge(combo, on="sample_id", how="left")

print("paired_combo shape:", paired_combo.shape)
display(paired_combo[["sample_id","subject_id","timepoint","Category","pct_high_MITO_ROS"]].head(10))

# Subject-level paired deltas: average across biopsies per subject per timepoint, then Post-Pre
paired_combo["resp_group"] = paired_combo["Category"].str.contains("Non_Remission").map({True:"Non_Remission", False:"Remission"})
subj_tp = paired_combo.groupby(["subject_id","resp_group","timepoint"], as_index=False)["pct_high_MITO_ROS"].mean()

pre = subj_tp[subj_tp["timepoint"]=="Pre"].set_index(["subject_id","resp_group"])
post = subj_tp[subj_tp["timepoint"]=="Post"].set_index(["subject_id","resp_group"])

paired_idx = pre.index.intersection(post.index)
d = (post.loc[paired_idx, ["pct_high_MITO_ROS"]] - pre.loc[paired_idx, ["pct_high_MITO_ROS"]]).reset_index()
d = d.rename(columns={"pct_high_MITO_ROS":"delta_pct_high_MITO_ROS"})

print("Paired subjects used:", len(d))
display(d.head())

summary = d.groupby("resp_group")["delta_pct_high_MITO_ROS"].agg(["median","mean","count"])
print("\nDelta(Post-Pre) pct_high_MITO_ROS by response:")
display(summary)


paired_combo shape: (110, 26)


sample_id subject_id timepoint          Category  pct_high_MITO_ROS
0  CID003352-2        UC1      Post  UC_Non_Remission          12.015504
1  CID003356-1        UC1       Pre  UC_Non_Remission          10.714286
2  CID003360-1        UC2      Post  UC_Non_Remission           5.806452
3  CID003361-1        UC2      Post  UC_Non_Remission          14.444444
4  CID003363-1        UC2       Pre  UC_Non_Remission          16.304348
5  CID003364-1        UC2       Pre  UC_Non_Remission           9.859155
6  CID003366-1        UC3      Post  UC_Non_Remission           9.170306
7  CID003367-2        UC3      Post  UC_Non_Remission          11.196911
8  CID003368-1        UC3      Post  UC_Non_Remission          13.302752
9  CID003370-1        UC3       Pre  UC_Non_Remission           7.339450

Paired subjects used: 32


subject_id     resp_group  delta_pct_high_MITO_ROS
0        CD1  Non_Remission                -0.333356
1       CD10      Remission                -1.738758
2       CD11  Non_Remission                -3.291285
3       CD12  Non_Remission                 2.374288
4       CD13      Remission                -2.665661


Delta(Post-Pre) pct_high_MITO_ROS by response:


median      mean  count
resp_group                              
Non_Remission -1.151986 -1.086653     18
Remission     -2.507425 -2.503437     14

In [20]:
import statsmodels.api as sm

# Build a subject-level delta table that includes:
# - delta_pct_high_MITO_ROS
# - deltas of inflammatory / TNFA mean scores
d2 = d.merge(
    deltas[["subject_id","resp_group",
            "mean_score_HALLMARK_INFLAMMATORY_RESPONSE",
            "mean_score_HALLMARK_TNFA_SIGNALING_VIA_NFKB"]],
    on=["subject_id","resp_group"],
    how="left"
)

# Rename for clarity
d2 = d2.rename(columns={
    "mean_score_HALLMARK_INFLAMMATORY_RESPONSE": "delta_inflammatory",
    "mean_score_HALLMARK_TNFA_SIGNALING_VIA_NFKB": "delta_tnfa"
})

# Encode resp_group as 0/1 (Non_Remission=1)
d2["nonremission"] = (d2["resp_group"] == "Non_Remission").astype(int)

print("Rows:", len(d2))
display(d2.head())

# Model: mito/ROS delta ~ nonremission + delta_inflammatory + delta_tnfa
X = d2[["nonremission", "delta_inflammatory", "delta_tnfa"]]
X = sm.add_constant(X)
y = d2["delta_pct_high_MITO_ROS"]

model = sm.OLS(y, X).fit()
print(model.summary())


Rows: 32


subject_id     resp_group  delta_pct_high_MITO_ROS  delta_inflammatory  \
0        CD1  Non_Remission                -0.333356            0.680385   
1       CD10      Remission                -1.738758           -0.219837   
2       CD11  Non_Remission                -3.291285           -0.128495   
3       CD12  Non_Remission                 2.374288            0.193617   
4       CD13      Remission                -2.665661           -0.348006   

   delta_tnfa  nonremission  
0    0.179057             1  
1   -1.439471             0  
2   -0.859057             1  
3    0.150592             1  
4   -0.913997             0

                               OLS Regression Results                              
Dep. Variable:     delta_pct_high_MITO_ROS   R-squared:                       0.130
Model:                                 OLS   Adj. R-squared:                  0.037
Method:                      Least Squares   F-statistic:                     1.393
Date:                     Sat, 07 Feb 2026   Prob (F-statistic):              0.265
Time:                             16:45:06   Log-Likelihood:                -79.832
No. Observations:                       32   AIC:                             167.7
Df Residuals:                           28   BIC:                             173.5
Df Model:                                3                                         
Covariance Type:                 nonrobust                                         
                         coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------

In [21]:
# Restrict to biopsies marked Non_Inflamed and redo subject-level deltas
df_nf = paired_combo[paired_combo["Inflammation"] == "Non_Inflamed"].copy()
print("Non_Inflamed biopsies:", len(df_nf), " unique subjects:", df_nf["subject_id"].nunique())

subj_tp_nf = df_nf.groupby(["subject_id","resp_group","timepoint"], as_index=False)["pct_high_MITO_ROS"].mean()
pre_nf = subj_tp_nf[subj_tp_nf["timepoint"]=="Pre"].set_index(["subject_id","resp_group"])
post_nf = subj_tp_nf[subj_tp_nf["timepoint"]=="Post"].set_index(["subject_id","resp_group"])

paired_idx_nf = pre_nf.index.intersection(post_nf.index)
d_nf = (post_nf.loc[paired_idx_nf, ["pct_high_MITO_ROS"]] - pre_nf.loc[paired_idx_nf, ["pct_high_MITO_ROS"]]).reset_index()
d_nf = d_nf.rename(columns={"pct_high_MITO_ROS":"delta_pct_high_MITO_ROS"})

print("Paired subjects (Non_Inflamed only):", len(d_nf))
summary_nf = d_nf.groupby("resp_group")["delta_pct_high_MITO_ROS"].agg(["median","mean","count"])
print("\nDelta(Post-Pre) pct_high_MITO_ROS in Non_Inflamed biopsies:")
display(summary_nf)


Non_Inflamed biopsies: 37  unique subjects: 19
Paired subjects (Non_Inflamed only): 4

Delta(Post-Pre) pct_high_MITO_ROS in Non_Inflamed biopsies:


median      mean  count
resp_group                              
Non_Remission -2.193828 -2.193828      2
Remission     -2.940007 -2.940007      2

In [22]:
import statsmodels.api as sm
import pandas as pd

# For each subject/timepoint: average pct_high_MITO_ROS and compute fraction inflamed
subj_tp2 = paired_combo.groupby(["subject_id","resp_group","timepoint"], as_index=False).agg(
    pct_high_MITO_ROS_mean=("pct_high_MITO_ROS","mean"),
    frac_inflamed=("Inflammation", lambda x: (x=="Inflamed").mean()),
    n_biopsies=("sample_id","nunique")
)

pre2 = subj_tp2[subj_tp2["timepoint"]=="Pre"].set_index(["subject_id","resp_group"])
post2 = subj_tp2[subj_tp2["timepoint"]=="Post"].set_index(["subject_id","resp_group"])
paired_idx2 = pre2.index.intersection(post2.index)

d3 = pd.DataFrame({
    "delta_pct_high_MITO_ROS": (post2.loc[paired_idx2,"pct_high_MITO_ROS_mean"] - pre2.loc[paired_idx2,"pct_high_MITO_ROS_mean"]),
    "delta_frac_inflamed": (post2.loc[paired_idx2,"frac_inflamed"] - pre2.loc[paired_idx2,"frac_inflamed"]),
    "n_biopsies_pre": pre2.loc[paired_idx2,"n_biopsies"],
    "n_biopsies_post": post2.loc[paired_idx2,"n_biopsies"],
}).reset_index()

d3["nonremission"] = (d3["resp_group"]=="Non_Remission").astype(int)

print("Paired subjects:", len(d3))
display(d3.head())

# Regression: mito/ROS delta ~ nonremission + delta_frac_inflamed
X = sm.add_constant(d3[["nonremission","delta_frac_inflamed"]])
y = d3["delta_pct_high_MITO_ROS"]
model2 = sm.OLS(y, X).fit()
print(model2.summary())


Paired subjects: 32


subject_id     resp_group  delta_pct_high_MITO_ROS  delta_frac_inflamed  \
0        CD1  Non_Remission                -0.333356                 0.00   
1       CD10      Remission                -1.738758                -1.00   
2       CD11  Non_Remission                -3.291285                 0.00   
3       CD12  Non_Remission                 2.374288                 0.00   
4       CD13      Remission                -2.665661                -0.25   

   n_biopsies_pre  n_biopsies_post  nonremission  
0               1                1             1  
1               2                2             0  
2               2                2             1  
3               1                1             1  
4               4                4             0

                               OLS Regression Results                              
Dep. Variable:     delta_pct_high_MITO_ROS   R-squared:                       0.060
Model:                                 OLS   Adj. R-squared:                 -0.004
Method:                      Least Squares   F-statistic:                    0.9309
Date:                     Sat, 07 Feb 2026   Prob (F-statistic):              0.406
Time:                             16:47:11   Log-Likelihood:                -81.063
No. Observations:                       32   AIC:                             168.1
Df Residuals:                           29   BIC:                             172.5
Df Model:                                2                                         
Covariance Type:                 nonrobust                                         
                          coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------

## Today’s findings (TAURUS discovery setup)

- Loaded TAURUS myeloid `.h5ad`: 30,858 cells; 216 biopsies (`sample_id`).
- Paired list: 110 biopsies overlap; 32 subjects have both Pre & Post.
- `score_modules` runs in-place and adds `score_HALLMARK_*` columns.
- State-high per biopsy: `assign_state_high(... q=0.8)`; combined state `MITO_ROS = OXPHOS_high & ROS_high`.
- Key result (paired subjects): median Δ(Post−Pre) %MITO_ROS_high is more negative in Remission than Non_Remission; regression adjusting for Δ inflamed fraction gives nonremission coef ≈ +2.26 (trend).
